content/learning-paths/mobile-graphics-and-gaming/android_halide/fusion.md

```diff
@@ -0,0 +1,199 @@
+---
+# User change
+title: "Demonstrating Operation Fusion"
+
+weight: 4
+
+layout: "learningpathall"
+---
+
+## Objective
+In this section, you’ll learn about operation fusion, a powerful performance optimization technique offered by Halide. We’ll explore how combining multiple processing stages into a single fused operation reduces memory usage, decreases scheduling overhead, and significantly enhances performance. You’ll see when and why to apply operation fusion, analyze the performance of a baseline pipeline, identify bottlenecks, and then leverage Halide’s scheduling constructs like compute_at, store_at, and fuse to optimize and accelerate your image-processing applications.
+
+## What is Operation Fusion?
+Operation fusion (also known as operator fusion or kernel fusion) is a technique used in high-performance computing, especially in image and signal processing pipelines, where multiple computational steps (operations) are combined into a single processing stage. Instead of computing and storing intermediate results separately, fused operations perform calculations in one continuous pass, reducing redundant memory operations and improving efficiency.
+
+## How Fusion Reduces Memory Bandwidth and Scheduling Overhead
+Every individual stage in a processing pipeline typically reads input data, computes intermediate results, writes these results back to memory, and then the next stage again reads this intermediate data. This repeated read-write cycle introduces significant overhead, particularly in memory-intensive applications like image processing. Operation fusion dramatically reduces this overhead by:
+1. Reducing memory accesses. Intermediate results stay in CPU registers or caches rather than being repeatedly written to and read from main memory.
+2. Improving cache utilization. Data is accessed in a contiguous manner, improving CPU cache efficiency.
+3. Reducing scheduling overhead. By executing multiple operations in a single pass, scheduling complexity and overhead are minimized.
+
+## When to Use Operation Fusion
+Operation fusion is most beneficial in scenarios that involve multiple sequential operations or transformations performed on large datasets, particularly when these intermediate results are large or costly to recompute. Typical situations include:
+* Image filtering pipelines (blur, sharpen, threshold sequences)
```

Comment:
In general, blur and spatial filters are not immediately good for fusion. Depending on the filter and pipeline structure, it would be better with tiling as the fusion requires more compute amount to get the intermediate values on the fly where we compute the same result multiple times. So, there is a trade-offs of memory locality and compute redundancy. For example, in case the kernel size of the filter is large, or multiple sequence of small convolution (5 layers of 3x3).

What fusion improves the most is a element-wise operation where we do not refer to the value of neighbour pixels unlike spatial filters.